thesis/03-implementation.tex

```diff
@@ -83,23 +83,491 @@ \subsubsection{Backend}
+
+\medskip
+The frontend is built and deployed on \tech{Netlify} with no additional configuration requirements. This configuration-less nature of the deployment allows switching hosting providers easily should a need arise.
+
+
+
+
 % Instrukcja postawienia środowiska deweloperskiego - jeśli potrzeba wypełniacza %
 
-\subsection{Quality assurance}
-% Quality Assurance: czy mamy testy jednostkowe? Czy mamy inne testy automatyczne? Jakich bibliotek używacie? %
 
 \subsection{Problems encountered}
 % Z jakimi problemami technicznymi sie borykaliście? Jeśli nie opisaliście ich w dokumentacji procesowej. %
+% Collissions
+\subsubsection{Collision detection and handling}
+\subsubsection{Damping force}
+\subsubsection{Paper stiffness}
+\subsubsection{Unnatural guide encoding}
+
+\subsubsection{Mesh rendering hazard}
+
+When folding origami models, two faces lying flat on each other are a common occurrence.
+
+The Solving algorithm is not sophisticated enough, to have a sense of sheet thickness. Therefore it places the folded faces close to each other, on the verge of numerical error. The rendering engine cannot distinguish which face is on top and which one is below, resulting in face permeation and color leakage demonstrated below.
+
+\begin{figure}[H]
+	\caption{An example of a mesh rendering hazard}
+  \centering
+    \includegraphics[width=0.6\textwidth]{assets/3-rendering-hazard.png}
+\end{figure}
+
+Unfortunately, the issue has not been resolved in the current iteration. Some potential solutions include:
+\begin{itemize}
+	\item preventing face collision during solving,
+	\item detecting face interpenetration after solving, and marking which one is on top.
+\end{itemize}
+
+\subsubsection{Large web application bundle size}
+
+The frontend application uses multiple feature-rich libraries of significant file size footprint such as \tech{Three.js} or \tech{Material-UI}. The whole bundle size was $ 6.81\mathrm{MB} $ resulting in a non-ideal download and parsing times of several seconds.
+
+\medskip
+To analyze the size of all bundle components, \tech{BundleAnalyzerPlugin} for \tech{Webpack} was used. 
+
+\medskip
+
+\begin{figure}[H]
+	\caption{The output of the analysis is in the form of an interactive website that allows drilling down and identifying potential areas of improvement.}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-bundle.png}
+\end{figure}
+
+To optimize the application size several techniques have been used:
+
+\begin{description}
+	\item[tree-shaking] - dead code elimination technique
+	\item[minimizing] - unnecessary character removal
+	\item[chunking] - code splitting and on-demand chunk loading
+\end{description}
+
+The minimizing was done using the \tech{Terser} plugin. Chunk splitting and tree-shaking feature are included, but not enabled by default in \tech{Webpack}.
+
+\begin{table}[H]
+\caption{Incorporating optimizations reduced the bundle size massively. Cutting the load time by a large factor.}
```

Comment:
These sentences sound like they don't like to be separated.

"Incorporating optimizations reduced the bundle size massively, cutting the load time by a large factor"